[2.15] Updates to CAPI PnP documentation and userdata security guidelines (#1271)
rancher/rancher-product-docs: versions/v2.15/modules/en/pages/cluster-deployment/configuration/capi-infrastructure-providers.adoc

Reviewer: Please update the date.

:url-CAPI-docs: https://turtles.docs.rancher.com/turtles/stable/en/index.html
:url-CAPIProvider: https://turtles.docs.rancher.com/turtles/stable/en/reference/capiprovider.html
:xref-enable-experimental-features: xref:how-to-guides/advanced-user-guides/enable-experimental-features/index.adoc
:xref-user-data-security: xref:reference-guides/rancher-security/rancher-security-best-practices.adoc#_restrict_access_to_user-data
:xref-cloud-credentials: xref:reference-guides/user-settings/manage-cloud-credentials.adoc
endif::[]

[#_overview]
== Overview

Rancher can provision RKE2 clusters using native CAPI infrastructure providers, such as https://cluster-api-aws.sigs.k8s.io/[CAPA] (Cluster API Provider AWS) and https://github.com/kubernetes-sigs/cluster-api-provider-vsphere[CAPV] (Cluster API Provider vSphere).

Standard RKE2 provisioning relies on Rancher's internal bootstrap and control plane providers alongside Rancher Node Drivers (via `rancher/machine`) as the infrastructure provider. This new mode allows you to substitute Rancher Node Drivers with native CAPI infrastructure providers while retaining Rancher’s bootstrap and control plane logic.

> This guide provides simple examples for this provisioning mode using CAPA and CAPV. Refer to the documentation of each provider for more details on available options and adapt these examples to your needs.

Reviewer: ```suggestion
This guide provides simple examples for this provisioning mode using CAPA and CAPV. Rancher Turtles is the supported method for installing CAPA and CAPV. Refer to the documentation of each provider for more details on available options and adapt these examples to your needs.
```

[NOTE]
====
Provisioning with a native CAPI infrastructure provider and Rancher as a bootstrap and control plane provider is distinct from using {url-CAPI-docs}[{turtles-product-name}] and the https://caprke2.docs.rancher.com/[CAPRKE2] provider to provision a RKE2 cluster and subsequently import it into Rancher.
====

[CAUTION]
====

> Standard users by default have permissions to create CAPI infrastructure objects in the `fleet-default` namespace.

Reviewer: ```suggestion
By default, users assigned the built-in Rancher `user` or `clusters-create` roles have permissions to create CAPI infrastructure objects in the `fleet-default` namespace.
```

> Infrastructure provider identity objects that target `fleet-default` can be used by any such user to provision cluster infrastructure through objects such `AWSCluster` or `VsphereCluster`.

Reviewer: ```suggestion
Infrastructure provider identity objects, such as `AWSClusterStaticIdentity` or `VSphereClusterIdentity`, that target the `fleet-default` namespace can be used by any such user to provision cluster infrastructure through objects such `AWSCluster` or `VsphereCluster`.
```

. Install Rancher.
. Install a CAPI infrastructure provider, either CAPA or CAPV.

> . Set-up an identity resource for the provider, or use an identity resource mirrored from a cloud credential.

Reviewer: ```suggestion
. Set-up an identity resource for the provider, or use an identity resource mirrored from a cloud credential. Automatic mirroring is currently supported only for CAPA's AWSClusterStaticIdentity.
```

[#_capa]
=== CAPA

First, configure IAM as required by CAPA. Different permissions are needed by the CAPA controller in the upstream cluster and by the downstream cluster to enable the Kubernetes AWS Cloud Provider.

To do this, CAPA provides the `clusterawsadm` tool to generate and apply the required objects. Refer to the CAPA manual for more https://cluster-api-aws.sigs.k8s.io/topics/using-clusterawsadm-to-fulfill-prerequisites[details].

Then, configure the provider identity in the upstream cluster so that the CAPA provider can create resources on AWS.

To do this, create a {xref-cloud-credentials}[cloud credential] for AWS. Rancher creates the corresponding `AWSClusterStaticIdentity` with the same name, if the CAPA provider is installed.

[CAUTION]
====

> Any user with permission to create `AWSCluster` objects in the `fleet-default` namespace is able to use manually created CAPA identity objects that target this namespace. This includes Rancher standard users, who have this permission by default.

Reviewer: ```suggestion
Any user with permission to create `AWSCluster` objects in the `fleet-default` namespace is able to use manually created CAPA identity objects that target this namespace. By default, users assigned the built-in Rancher `user` or `clusters-create` roles have this permission.
```

> However, identities created by Rancher from cloud credentials can only be referenced in `AWSCluster` objects created in `fleet-default` by users with read access to the cloud credential.

Reviewer: ```suggestion
However, identity resources automatically mirrored by Rancher from cloud credentials can only be referenced by `AWSCluster` objects created by users who have read access to the underlying cloud credential.
```

> Currently Rancher only creates identity objects from cloud credentials for CAPA.

Reviewer: ```suggestion
Currently Rancher only creates `AWSClusterStaticIdentity` objects from cloud credentials for CAPA.
```

            - --configure-cloud-routes=false
            - --v=5
            - --cloud-provider=aws
          tolerations:
                - key: node.cloudprovider.kubernetes.io/uninitialized
                  value: "true"
                  effect: NoSchedule
                - key: node-role.kubernetes.io/master
                  effect: NoSchedule
                - key: node-role.kubernetes.io/control-plane
                  effect: NoSchedule
                - key: node-role.kubernetes.io/etcd
                  effect: NoExecute
----

[#_capv]
=== CAPV

First, configure the provider identity in the upstream cluster so that the CAPV provider can create resources on your vSphere server. Refer to the manual for all identity https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/v1.15.2/docs/identity_management.md[options], and for general vSphere https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/v1.15.2/docs/getting_started.md[requirements].

[CAUTION]
====

> Any user with permission to create `VsphereCluster` objects in the `fleet-default` namespace is able to use manually created identities such as the one in the example below to provision CAPV clusters. This includes Rancher standard users, who have this permission by default.

Reviewer: ```suggestion
Any user with permission to create `VsphereCluster` objects in the `fleet-default` namespace is able to use manually created identities such as the one in the example below to provision CAPV clusters. By default, users assigned the built-in Rancher `user` or `clusters-create` roles have this permission.
```

Author: I've updated these with the terminology from https://ranchermanager.docs.rancher.com/how-to-guides/new-user-guides/authentication-permissions-and-global-configuration/manage-role-based-access-control-rbac/global-permissions.

  name: capv-lab
  namespace: fleet-default
spec:

>   kubernetesVersion: v1.35.1+rke2r1

Reviewer: Should it be kept and changed to something like: 
```
 kubernetesVersion: <RKE2 version, e.g. v1.35.1+rke2r1>
```

Author: Well, the version was there just to provide some context, since I removed most of the fields to emphasize how to add the userdata without adding a full provisioning cluster configuration. I didn't want to keep a fixed version like before, so I thought it was easier to just remove it.
